mkimg: support raw images and misc improvements

LekKit opened the pull request:
This PR is an attempt to adapt `mkimg` for different VMs as well as real hardware hopefully (Unmatched, VisionFive).
Any objections are appreciated, so far it's a mix of changes and I may split it into several commits later or remove unwanted stuff.

- Determine whether raw image should be generated based on filename extension
- Replace hardcoded root=/dev/vda1 with root=UUID=..., this allows different configurations like booting off NVMe or MMC, works on RVVM and likely on physical hardware too
- This allows to flash the resulting distro onto physical storage and boot on real board. Tested on my VisionFive 2, although the kernel hangs later on due to JH7110 not yet being properly mainlined. Someone with Unmatched should have a look at this I guess.
- Show both initramfs' in boot menu and prefer non-fallback one, this fixes slow boot times, and doesn't hang on boards with low RAM (256M)*
- Allow to skip firmware build using the -d option, to speed up image build if you don't need the firmware
- Use --depth 1 to significantly speed up git clone
- Add U-Boot patch to support NVMe drives, this allows to run the resulting firmware on RVVM (May be removed if anyone has objections)**
- Update OpenSBI to v1.2

* I am aware of #4 but I believe if something is broken in generic initramfs, it should be fixed directly. The fallback initramfs is too bloated and outright broken on some platforms. Moreover, I couldn't reproduce the original issue so it might have been fixed already?
** This patch hopefully will be accepted in U-Boot mainline at some point because lacking NVMe in the config affects QEMU users too.

![image](https://github.com/CoelacanthusHex/archriscv-scriptlet/assets/50500857/8b3932ec-f307-460a-b599-548702d11dac)

LekKit committed:
mkimg: Support raw images, misc improvements

- Determine whether raw image should be generated based on filename extension
- Replace hardcoded root=/dev/vda1 with root=UUID=..., this allows different configurations like booting off NVMe or MMC, works on RVVM and likely on physical hardware too
- This allows to flash the resulting distro onto physical storage and boot on real board. Tested on my VisionFive 2, although the kernel hangs later on due to JH7110 not yet being properly mainlined
- Show both initramfs in boot menu and prefer non-fallback one, this fixes slow boot times, and works on boards with low RAM (256M)
- Allow to skip firmware build using the -d option, to speed up image build if you don't need the firmware
- Use --depth 1 to significantly speed up git clone
- Add U-Boot patch to support NVMe drives, this allows to run the resulting firmware on RVVM (May be removed if anyone has objections)
- Update OpenSBI to v1.2

CoelacanthusHex commented:
> This PR is an attempt to adapt `mkimg` for different VMs as well as real hardware hopefully (Unmatched, VisionFive). Any objections are appreciated, so far it's a mix of changes and I may split it into several commits later or remove unwanted stuff.
> 
>     * Determine whether raw image should be generated based on filename extension
> 
>     * Replace hardcoded root=/dev/vda1 with root=UUID=..., this allows different configurations like booting off NVMe or MMC, works on RVVM and likely on physical hardware too
> 
>     * This allows to flash the resulting distro onto physical storage and boot on real board. Tested on my VisionFive 2, although the kernel hangs later on due to JH7110 not yet being properly mainlined. Someone with Unmatched should have a look at this I guess.

Great. mkimg script is just a quick way to setup a qemu-system build instance originally, so we don't consider other cases at begin.

>     * Show both initramfs' in boot menu and prefer non-fallback one, this fixes slow boot times, and doesn't hang on boards with low RAM (256M)*

We prefer fallback one because we generate initramfs in a different environment compared to boot. It may lack some drivers to boot, and has happened before.

>     * Allow to skip firmware build using the -d option, to speed up image build if you don't need the firmware

Agreed.

>     * Use --depth 1 to significantly speed up git clone

Agreed. But it shouldn't be an issue because it only pull full repo when first use, and will reuse previous clone.

>     * Add U-Boot patch to support NVMe drives, this allows to run the resulting firmware on RVVM (May be removed if anyone has objections)**

Could you send this patch to uboot upstream? We prefer use upstream source. I think it's usefull for your patch maintain as well.

>     * Update OpenSBI to v1.2
> 
> 
>     * I am aware of [refractor(mkimg): using fallback initramfs directly #4](https://github.com/CoelacanthusHex/archriscv-scriptlet/pull/4) but I believe if something is broken in generic initramfs, it should be fixed directly. The fallback initramfs is too bloated and outright broken on some platforms. Moreover, I couldn't reproduce the original issue so it might have been fixed already?

As defination of non-fallback initramfs, it's used in the **SAME** environment as generating environment. If you change environment, you should use fallback one to boot first, and then re-generate non-fallback initramfs to use later. So it's not a bug of initramfs generator. It's just expected by design. Insert module is some kind hack. So I prefer to add two entries: non-fallback and fallback, but leave fallback default. Users can come back to non-default one after first to reduce overhead. And for first boot memory usage, I think only virtual machine like QEMU will has memory lower than 512MB. The minium board our kernel supports is Lichee RV Nezha, which has 512MB memory. But virtual machine's memory is mutable, so I think it's not a serious problem.

>       ** This patch hopefully will be accepted in U-Boot mainline at some point because lacking NVMe in the config affects QEMU users too.

Could you provide a patch link in uboot mail list?

> 
> ![image](https://user-images.githubusercontent.com/50500857/242870383-8b3932ec-f307-460a-b599-548702d11dac.png)

LekKit commented:
> Agreed. But it shouldn't be an issue because it only pull full repo when first use, and will reuse previous clone.

This could help in cases like running this in CI, or when a new user just wants to build this fresh / a newer firmware is released.

> Could you send this patch to uboot upstream? We prefer use upstream source. I think it's usefull for your patch maintain as well.

Yeah no problem. I will remove it from this PR for now.

> As defination of non-fallback initramfs, it's used in the **SAME** environment as generating environment. If you change environment, you should use fallback one to boot first, and then re-generate non-fallback initramfs to use later. So it's not a bug of initramfs generator. It's just expected by design. Insert module is some kind hack. So I prefer to add two entries: non-fallback and fallback, but leave fallback default. Users can come back to non-default one after first to reduce overhead. And for first boot memory usage, I think only virtual machine like QEMU will has memory lower than 512MB. The minium board our kernel supports is Lichee RV Nezha, which has 512MB memory. But virtual machine's memory is mutable, so I think it's not a serious problem.

Makes sense, thanks. The idea was to at least include the default initramfs in the boot menu and allow to pick fallback one if anything goes wrong.
Moreover, the basic set of drivers needed to get root device to work is pretty minimal yet fallback initramfs pulls a lot of unwanted stuff like GPU drivers and firmware that could be deferred to the point when we get the root device. It's a matter of preference I guess.

Theoretically makes sense to shoehorn some kind of deferred `mkinitcpio`? Like make the `/boot/initramfs-linux.img` a symlink to fallback one originally, then regenerate it on first boot.... Alright, sounds too hacky for a generic distro images

The fact is, bringing up fallback initramfs takes **a minute** for me on QEMU, which is kinda a lot

LekKit committed:
mkimg: Remove patch and fix typos

- Remove NVMe qemu board patch from U-Boot as using mainline stuff is preferred
- Fix typo in firmware build condition
- Prevent overwriting files when extracting the roots and omit reinstalling the kernel package if it's already there; Mainly to prevent screwing up existing installations by accident

LekKit committed:
mkimg: Prefer fallback initramfs

LekKit commented:
Using uncompressed initramfs with compressed modules inside makes for a **MUCH** faster initramfs boot stage (talk about 1 minute -> 1 second) since the kernel doesn't have to decompress the whole initramfs but only individual modules.
This also reduces the RAM bubbling a bit since the modules are kept compressed all the time, except those that are actually loaded. The kernel doesn't waste memory on initramfs decompression as well.
The actual size of the initramfs isn't affected much, primarily due to modules and firmware being already compressed so no point of doing it twice.

Question: Is such configuration desired by default? Because it seems very good for both end users and maintainers so I guess this wouldn't hurt as a stock config.

@fish4terrisa-MSDSM this could help improve boot times on your project, and also this script now may be used to build images for archriscv-term. Hope this helps.

LekKit commented:
> Question: Is such configuration desired by default? Because it seems very good for both end users and maintainers so I guess this wouldn't hurt as a stock config.

I guess this may be later reviewed in a separate PR.

Any objections on this one? U-Boot patch removed, using fallback initramfs by default, fixed a few nuances. I hope it is good to go.

CoelacanthusHex commented:
> > Question: Is such configuration desired by default? Because it seems very good for both end users and maintainers so I guess this wouldn't hurt as a stock config.
> 
> I guess this may be later reviewed in a separate PR.
> 
> Any objections on this one? U-Boot patch removed, using fallback initramfs by default, fixed a few nuances. I hope it is good to go.

I'm busy with China NCEE, so I will review this PR carefully later.

fish4terrisa-MSDSM commented:
> Using uncompressed initramfs with compressed modules inside makes for a **MUCH** faster initramfs boot stage (talk about 1 minute -> 1 second) since the kernel doesn't have to decompress the whole initramfs but only individual modules. This also reduces the RAM bubbling a bit since the modules are kept compressed all the time, except those that are actually loaded. The kernel doesn't waste memory on initramfs decompression as well. The actual size of the initramfs isn't affected much, primarily due to modules and firmware being already compressed so no point of doing it twice.
> 
> Question: Is such configuration desired by default? Because it seems very good for both end users and maintainers so I guess this wouldn't hurt as a stock config.
> 
> @fish4terrisa-MSDSM this could help improve boot times on your project, and also this script now may be used to build images for archriscv-term. Hope this helps.

Hmm...I will try to build RVVM and use this script to build Images in circleci workflow .However, as circleci don\`t support loop mount ,it will take me some time to use User mode linux to fix that.I may release a new image later.

fish4terrisa-MSDSM commented:
> We prefer fallback one because we generate initramfs in a different environment compared to boot. It may lack some drivers to boot, and has happened before.
I just build the image inside RVVM ,and the non-fallback one worked well.I think it may be qemu-user which made it generate wrong image, not mkinitcpio\`s problem .And i think it\`s better to build images for the same environment which runs it.Although the lacking of drivers may happen ,it\`s really rare.

LekKit commented:
Last thing to handle is probably some kind of dialog when the script encounters an already partitioned device.
I tried to `./mkimg /dev/mmcblk0` to directly flash my SD using it, but it resulted in rootfs being squeezed into the first (small) partition on it. Passing `echo o` into `fdisk` will forcefully repartition it, which is not nice either if you do this by mistake...

LekKit commented:
Any update on this?

CoelacanthusHex commented:
LGTM, I will test and merge it later.
